lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,234 @@
+"""
+Models used for robust grading.
+
+Robust grading allows student scores to be saved per-subsection independent
+of any changes that may occur to the course after the score is achieved.
+"""
+
+from django.db import models
+from model_utils.models import TimeStampedModel
+
+from coursewarehistoryextended.fields import UnsignedBigIntAutoField
+from opaque_keys.edx.locator import BlockUsageLocator
+from xmodule_django.models import CourseKeyField, UsageKeyField
+
+from hashlib import md5
+import json
+
+
+class VisibleBlocks(models.Model):
+    """
+    A django model used to track the state of a set of visible blocks under a given subsection at the time they are
+    used for grade calculation.
+
+    This state is represented using an array of serialized BlockRecords, stored in the blocks_json field. A
+    hash of this json array is used for lookup purposes.
+    """
+    _blocks_json = models.TextField(db_column="blocks_json")
+    hashed = models.CharField(max_length=32, primary_key=True)
+
+    def __unicode__(self):
+        """
+        String representation of this model.
+        """
+        return hashed
+
+    @classmethod
+    def create(cls, blocks):
+        """
+        Creates a new VisibleBlocks model. Argument 'blocks' should be an array of BlockRecords.
+        """
+        # Start by sorting the blocks to ensure equality regardless of how the blocks are ordered
+        sorted_blocks = sorted(
+            blocks,
+            key=lambda block: '{}{}{}'.format(block.block_key, block.max_score, block.weight)
+        )
+        _blocks_json = json.dumps(
+            [
+                block.to_dict()
+                for block in sorted_blocks
+            ]
+        )
+        hashed = md5(_blocks_json).hexdigest()
```

Comment:
The concern with md5 is not about random collisions, but engineered collisions.  Since external users create the courses, they do have the ability to engineer a collision.  Not that I expect edx.org course teams to try to bust our system, but Open edX should have a higher level of robustness.  There could be installations allowing arbitrary users to upload courses.  I'd recommend using sha256 instead.